keymanapp/keyman
# refactor(web): active element management in relation to OSK display control ✂️

Addresses aspects of #5026.

Follows #5633.

Fixes #5026.  (Finally!) This PR finally centralizes management of KMW's active element logic and fixes a few niche scenarios that went relatively undetected until now.

One of the worst aspects of "spaghetti code" in relation to the OSK is how entangled some of its display aspects have been with the web engine's active-element management.  While this PR doesn't get everything cleaned up perfectly for active-element tracking, its biggest success is that all logic where the two subsystems intersect has now been centralized within the setters of two properties.

While not actually refactoring OSK code in _this_ PR, it's prep work necessary to continue along the lines of the other ✂️
PRs.

## User Testing 

### SUITE_IN_APP:  Host app tests.

All behavior exhibited here should feel "normal".  These tests are solely to ensure that nothing got accidentally broken.

**Platform / keyboard mode combinations**
- **GROUP_IOS_APP**: iOS / Keyman app
- **GROUP_ANDROID_APP**: Android / System keyboard

**Suite tests**
- **TEST_NORMAL_LOAD**:  Ensure that the keyboard displays and works correctly when first loaded up.

    Be sure to test one or two keystrokes; ensure that the key produces the expected output.

- **TEST_ROTATE_P-TO-L**:

    1. With Keyman not active, load it in a portrait orientation.
    2. Once loaded, rotate the device to a landscape orientation.
        - On Android devices, you may need to press something like this for the rotation to occur:
        - The test "passes" if the OSK rotates properly.

- **TEST_ROTATE_L-TO-P**:

    1. With Keyman not active, load it in a landscape orientation.
    2. Once loaded, rotate the device to a portrait orientation.
        - On Android devices, you may need to press something like this for the rotation to occur:
        - The test "passes" if the OSK rotates properly.

### SUITE_DOM:  Site-based tests.

All behavior exhibited here should feel "normal".  These tests are solely to ensure that nothing got accidentally broken.

**OS / Browser combinations**
- **GROUP_WIN**: Windows / Firefox (or Chrome, but just one)
- **GROUP_ANDROID_DOM**: Android / Chrome
- **GROUP_IOS_DOM**: iOS (via Simulator) / Safari 
    - Note:  touches near the bottom of the screen may have some interference from Safari; it likes to display the navigation button bar and "shove" KMW's OSK up a bit.  This is expected and not a bug; the relocated OSK should function fine afterward.

For each of these, make use of the "Tests the new Attachment/Enablement API functionality" test page.

**Suite tests**
- **TEST_NORMAL_USE**:  Ensure that the keyboard displays and works correctly when first loaded up.

    1.  Click the "Create Inputs" button one time.
    2. Click / touch the new page element.  The OSK should display.
    3. Use the OSK to type one or two keys.  The keys should produce their expected output.
    4. Click / touch a blank area of the page.  The OSK should automatically hide.

- **TEST_ELEMENT_HOPPING**:

    1.  Click the "Create Inputs" button one time and the "Create Textarea" button one time.
    2. Click / touch the first of the new page elements.  The OSK should display.
    3. Click / touch the other page element.  The OSK should remain visible.
    4. Click / tough the first element again.  The OSK should remain visible.
    5. Click / touch a blank area of the page.  The OSK should automatically hide.

- **TEST_SPECIFIC_KEYBOARDS**:

    1.  Click the "Create Inputs" button three times.
    2. Click / touch `Dynamic area #0.`
    3. Click the element's "Set to Dzongkha" button.
    4. Change the keyboard to French.
    5. Click `Dynamic area #1`'s "Set to Dzongkha button.
    6. Click `Dynamic area #1`.  The Dzongkha keyboard should be displayed.
    7. Click `Dynamic area #2`.
    8. Change the keyboard to Lao.
    9. Click `Dynamic area #0.`  The French keyboard should be displayed.
    10. Click `Dynamic area #2.`  The Lao keyboard should be displayed.
    11. Click `Dynamic area #1.`  The Dzongkha keyboard should be displayed.
    12. Click `Dynamic area #0`'s "Clear Keyboard" button.
    13. Click `Dynamic area #0.`  The Lao keyboard should be displayed.

### SUITE_CJK:  Tests CJK keyboard functionality

**OS / Browser combinations**
- GROUP_MAC:  macOS / Firefox

Use the standard "Test unminified KeymanWeb" page.

**Suite tests**
- TEST_JAPANESE_TYPING:  Verify that a CJK keyboard still works.

    <details><summary>Test sequence</summary>

    1. Under "Add a keyboard by keyboard name", type `japanese`, then click Add.
    2. Select the Japanese keyboard.  The OSK should change shape notably:

    <img width="347" alt="image" src="https://user-images.githubusercontent.com/25213402/131969825-d6705f1a-abef-4059-842f-e387e57ae67b.png">

    3. Type `a`.  A "picker" displaying a few options should display:

    <img width="460" alt="image" src="https://user-images.githubusercontent.com/25213402/131969902-1be50127-d09d-40be-9fce-b3b6b8840cf0.png">

    4. Type `2`.  The second option should replace the context.

    </details>

- TEST_JAPANESE_FOCUS:  Verify that a CJK keyboard is shown and hidden properly without interfering with a page's UX.

    1. Select the page's textarea element.  The OSK's title bar should appear (as with step 2 for the `_TYPING` test).
    2. Click a blank area on the page.  The OSK should automatically hide and the caret should disappear from the textarea.
        -  Just in case, "caret":  the blinking vertical line segment that lets you know where new characters will appear in the context.
    3. Select the page's input field (the second editable control).  The OSK should reappear under the newly-selected control.
    4. Select the page's textarea element again.  The OSK should remain visible, and the caret should move from the input control to the textarea control.

-----

So, fun fact... the `TEST_JAPANESE_FOCUS` test above actually fixes an issue that currently exists both on `master` _and_ on `stable_14.0` that had gone unnoticed until this PR's development.
